### PR TITLE
Revamp the README documentation of the rules.

### PR DESCRIPTION
**Pull request by stefanbucur**
To do so, this change bumps up the python_rules dependency, which simplifies the import of PIP rules. It also adds a new RE2 example, which showcases a "real" API and demonstrates the fuzz test's ability to discover new coverage.

### CODE REVIEW

File: fuzzing/requirements.txt [hunk 1]
```diff
@@ -1 +1,4 @@
-absl-py>=0.9.0
+# Python requirements for the tools supporting the fuzzing rules. These are
+# installed automatically through the WORKSPACE configuration.
+
+absl-py>=0.11.0
```
Comment (inferno-chromium) [inline fuzzing/requirements.txt:4]: nit: >= can be risky since you never know a new version breaks stuff, but upto you, we freeze version in other places.

File: README.md [hunk 2]
```diff
@@ -103,3 +146,12 @@ A fuzzing engine launcher script receives configuration through the following en
...
 | `FUZZER_ARTIFACTS_DIR`     | A path under `FUZZER_OUTPUT_ROOT` where generated crashes and other relevant artifacts should be stored. |
+
+## Rule reference
+
+* [`cc_fuzz_test`](/docs/cc-fuzzing-rules.md#cc_fuzz_test)
+* [`cc_fuzzing_engine`](/docs/cc-fuzzing-rules.md#cc_fuzzing_engine)
+
+## Disclaimer
+
+This is not an officially supported Google product.
```
Comment (inferno-chromium) [inline README.md:157]: Do we want this disclaimer, this will be supported, and is an important part of our fuzzing portolio ?

File: README.md [hunk 3]
```diff
@@ -1,91 +1,134 @@
...
+### Configuring the .bazelrc file
+
+To make sure the fuzz tests are built with the correct instrumentation flags for each engine / instrumentation supported, we recommend using the configurations defined in this repository's [`.bazelrc` file](/.bazelrc), which you can copy and paste in your own `.bazelrc` file.
+
+Currently, the following configurations are available, based on the fuzzing engines defined in this repository:
+
+| Configuration             | Fuzzing engine | Instrumentation          |
+|---------------------------|----------------|--------------------------|
+| `--config=asan-fuzzer`    | libFuzzer      | Address Sanitizer (ASAN) |
+| `--config=msan-fuzzer`    | libFuzzer      | Address Sanitizer (ASAN) |
```
Comment (inferno-chromium) [inline README.md:128]: This needs to Memory Sanitizer (MSAN)

File: README.md [hunk 4]
```diff
@@ -1,91 +1,134 @@
...
+| `--config=asan-honggfuzz` | Honggfuzz      | Memory Sanitizer (MSAN)  |
```
Comment (inferno-chromium) [inline README.md:129]: This needs to Address Sanitizer (ASAN)

File: README.md [hunk 5]
```diff
@@ -1,91 +1,134 @@
...
-$ bazel run fuzz_test_run --config=asan-libfuzzer -- --timeout_secs=20
+$ bazel run -c opt --config=asan-honggfuzz //examples:re2_fuzz_test_run
 ```
 
-If you only want to run the regression test on the corpus, set `--regression`:
+The `<name>_run` target accepts a number of engine-agnostic flags. For example, the following command runs the fuzz test with an execution timeout and on a clean slate (removing any previously generated tests). Note the extra `--` separator between Bazel's own flags and the launcher flags:
 
 ```sh
-$ bazel run fuzz_test_run --config=asan-libfuzzer -- --regression=True
+$ bazel run -c opt --config=asan-libfuzzer //examples:re2_fuzz_test_run -- --clean --timeout_secs=30
```
Comment (inferno-chromium) [inline README.md:87]: nit: Maybe break to new line \
-- --clean --timeout_secs30
to make it more obvious ?

File: README.md [hunk 6]
```diff
@@ -1,91 +1,134 @@
...
 
-- `--config=asan-libfuzzer` builds the fuzz target in [libFuzzer](https://llvm.org/docs/LibFuzzer.html) mode, with [Address Sanitizer (ASAN)](https://clang.llvm.org/docs/AddressSanitizer.html) instrumentation.
-- `--config=msan-libfuzzer` builds the fuzz target in libFuzzer mode, with [Memory Sanitizer (MSAN)](https://clang.llvm.org/docs/MemorySanitizer.html) instrumentation.
-- `--config=asan-honggfuzz` builds the fuzz target using [Honggfuzz](https://github.com/google/honggfuzz), with ASAN instrumentation.
+The rule library currently provides support for C++ fuzz tests. Support for additional languages may be added in the future.
 
-The rest of the documentation assumes the build configuration options are accessible through these names.
+## Prerequisites
+
+C++ fuzz tests require a Clang compiler. The libFuzzer engine requires at least Clang 5.0.
```
Comment (inferno-chromium) [inline README.md:17]: i think 5.0 might be too old, i remember something like 9.0, can update after checking with kcc or dynamic-tools.

**Review by inferno-chromium — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -1,91 +1,134 @@
 # Bazel Rules for Fuzz Tests
 
-This repository contains [Bazel](https://bazel.build/)
-[Starlark extensions](https://docs.bazel.build/versions/master/skylark/concepts.html)
-for defining fuzz tests in Bazel projects. 
+## Overview
 
-**This is not an officially supported Google product.**
+This repository contains [Bazel](https://bazel.build/) [Starlark extensions](https://docs.bazel.build/versions/master/skylark/concepts.html) for defining fuzz tests in Bazel projects.
```
Comment (oliverchang) [inline README.md:5]: nit: I feel overall that this top-level README is a bit too wordy. Perhaps would be better to move parts of this to links to more detailed places? 

Ideally the README should just be a few lines showing a very simple example on how to get started very quickly. 

For more detailed information, might be better to move these other sections to sub-pages, especially sections like "Defining fuzzing engines" which should be less common but take up a lot of space here.

**Review by oliverchang — COMMENTED** [review]
The content looks great overall, but just a comment on possibly splitting this up. I think we should make the top-level README as short and approachable as possible, and break up the more detailed sections into other files. WDYT?

File: README.md (hunk 5 above)
Comment (stefanbucur) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (stefanbucur) [reply README.md]: Great catch, thanks!

File: README.md (hunk 4 above)
Comment (stefanbucur) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (stefanbucur) [reply README.md:5]: I am open to structuring this documentation differently. The reason I went with this approach is that this seems to be the prevailing form of documentation for Bazel rules. See for example:

* https://github.com/bazelbuild/rules_python
* https://github.com/bazelbuild/rules_go
* https://github.com/bazelbuild/rules_kotlin

And there are others like that, too.

I don't feel too strongly about this, so let me know if you prefer to take a different approach here.

File: README.md (hunk 6 above)
Comment (stefanbucur) [reply README.md]: Good catch. LibFuzzer's [documentation](https://llvm.org/docs/LibFuzzer.html#fuzzer-usage) actually mentions Clang 6.0 as being the minimum version:

> Recent versions of Clang (starting from 6.0) include libFuzzer, and no extra installation is necessary.

Updated to 6.0, but let me know if there are specific libFuzzer features that we should require and that are available in more recent versions.

**Review by stefanbucur — COMMENTED** [review]
PTAL.

**Comment by stefanbucur** [comment]
@inferno-chromium @oliverchang When you get the chance, please let me know if my latest comments address your concerns. Thanks a lot for your time reviewing this!

File: README.md (hunk 6 above)
Comment (inferno-chromium) [reply README.md]: ok fine then.

**Review by oliverchang — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (inferno-chromium) [reply README.md:158]: As long as we have someone maintaining it, fixing bugs, then it is not needed. I havent seen it used often, e.g. https://github.com/dvyukov/go-fuzz
Upto you. Adding it decreases credibility slightly as well.

**Comment by inferno-chromium** [comment]
LGTM

File: README.md (hunk 2 above)
Comment (stefanbucur) [reply README.md:158]: Good point. I'm also seeing now that other Bazel rules repositories don't mention that either, so I'm removing it from here as well.
